### ISSUE TITLE
Allow or document a way for FederatedTypeResolvers to provide GraphQLErrors per batch element

### ISSUE THREAD

**Issue by roookeee** [ExpediaGroup/graphql-kotlin#1563]
**Is your feature request related to a problem? Please describe.**
`FederatedTypeResolver<T>` is specified to provide a `List<T?>` which makes the idiomatic approach (specifying `T` as the actual type and not `Any`) unable to return a `GraphQLError` on a per element basis. A `FederatedTypeResolver<T>` can only provide a `GraphQLError` when throwing an exception which renders the whole batch invalid, making every response value `null` and generating a generic, framework defined error message. This is undesireable as a `FederatedTypeResolver<T>` could be used on a nullable receiver which would still like an error list while receiving the partially resolvable result. Furthermore a custom error message might be desireable.

**Describe the solution you'd like**
Change the API of `FederatedTypeResolver<T>` to allow for returning a `GraphQLError` as well, maybe with a wrapper object. If this is undesireable, the documentation should be updated on how to provide a `GraphQLError` on a per element basis (with setting `T` as `Any`) as we found out about the undocumented / unofficial support for returning a `GraphQLError` by inspecting the code. We usually don't want to rely on undocumented and thus potentially unintentionally supported solutions.

**Describe alternatives you've considered**
See above.

**Comment by samuelAndalon** [comment]
Hello, have you looked into 
https://opensource.expediagroup.com/graphql-kotlin/docs/schema-generator/execution/exceptions/#returning-data-and-errors ?

TLDR; you can return `GraphQLError`.

also FYI in `7.x.x` we are changing the signature of `FederatedTypeResolver` to return granular elements instead of lists, that might be beneficial for this use case, along with the usage of DataLoaders with the addition of CompletableFuture support

**Comment by roookeee** [comment]
Can I return a `DataFetcherResult<T>` inside a `FederatedTypeResolver`? I can't seem to find any documentation about the support. Hence this ticket: it's not documented what return types are supported: `Try<T>`, `DataFetcherResult<T>`, `Optional<T>` etc.

And yes, this seems like a good candidate to integrate into the 7.x.x. changes!

**Comment by samuelAndalon** [comment]
@roookeee created (#1568), will be included in next major release (we still have some tasks in the milestone).

what you would need to do is to just throw an exception when resolving a federated type, the corresponding value element in the array response will be `null` and `errors` section of response will be appended with a `GraphQLError`

**Comment by roookeee** [comment]
Thank you @samuelAndalon. Doesn't the current implementation break batching as a whole when not using `DataLoaders` in a `FederatedTypeResolver` ? Furthermore the `GraphQLErrorHandler` (which can be user overriden) should be used to handle the thrown exception to get consistent error handler across "normal" and "federation" queries.

Thank you for your work!

**Comment by samuelAndalon** [comment]
>Doesn't the current of https://github.com/ExpediaGroup/graphql-kotlin/pull/1568 implementation break batching as a whole when not using DataLoaders in a FederatedTypeResolver

The idea behind the change was to also achieve consistency across other subgraphs implementations, Apollo Server for example handles each element in the representations query in a granular way.

having said that, the usage of data loaders is almost a MUST when resolving entities.
